test_conformance/math_brute_force/unary_u_half.cpp

```diff
@@ -0,0 +1,238 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "common.h"
+#include "function_list.h"
+#include "test_functions.h"
+#include "utility.h"
+#include "reference_math.h"
+
+#include <cstring>
+#include <cinttypes>
+
+////////////////////////////////////////////////////////////////////////////////
+static cl_int BuildKernel_HalfFn(cl_uint job_id, cl_uint thread_id UNUSED,
+                                 void *p)
+{
+    BuildKernelInfo &info = *(BuildKernelInfo *)p;
+    auto generator = [](const std::string &kernel_name, const char *builtin,
+                        cl_uint vector_size_index) {
+        return GetUnaryKernel(kernel_name, builtin, ParameterType::Half,
+                              ParameterType::UShort, vector_size_index);
+    };
+    return BuildKernels(info, job_id, generator);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+int TestFunc_Half_UShort(const Func *f, MTdata d, bool relaxedMode)
+{
+    int error;
+    Programs programs;
+    KernelMatrix kernels;
+    const unsigned thread_id = 0; // Test is currently not multithreaded.
+    float maxError = 0.0f;
+    int ftz = f->ftz || gForceFTZ || 0 == (CL_FP_DENORM & gHalfCapabilities);
+    float maxErrorVal = 0.0f;
+    size_t bufferSize = BUFFER_SIZE;
+    size_t bufferElements = bufferSize / sizeof(cl_half);
+    uint64_t step = getTestStep(sizeof(cl_half), BUFFER_SIZE);
+    int scale = (int)((1ULL << 32) / (16 * bufferElements) + 1);
+    logFunctionInfo(f->name, sizeof(cl_half), relaxedMode);
+    const char *name = f->name;
+    float half_ulps = f->half_ulps;
+    if (gWimpyMode)
+    {
+        step = (1ULL << 32) * gWimpyReductionFactor / (512);
+    }
+
+    // Init the kernels
+    BuildKernelInfo build_info = { 1, kernels, programs, f->nameInCode };
+    if ((error = ThreadPool_Do(BuildKernel_HalfFn,
+                               gMaxVectorSizeIndex - gMinVectorSizeIndex,
+                               &build_info)))
+    {
+        return error;
+    }
+
+    for (uint64_t i = 0; i < (1ULL << 32); i += step)
+    {
+        // Init input array
+        cl_ushort *p = (cl_ushort *)gIn;
+        if (gWimpyMode)
+        {
+            for (size_t j = 0; j < bufferElements; j++) p[j] = i + j * scale;
+        }
+        else
+        {
+            for (size_t j = 0; j < bufferElements; j++) p[j] = (uint16_t)i + j;
+        }
+
+        if ((error = clEnqueueWriteBuffer(gQueue, gInBuffer, CL_FALSE, 0,
+                                          bufferSize, gIn, 0, NULL, NULL)))
+        {
+            vlog_error("\n*** Error %d in clEnqueueWriteBuffer ***\n", error);
+            return error;
+        }
+
+        // write garbage into output arrays
+        for (auto j = gMinVectorSizeIndex; j < gMaxVectorSizeIndex; j++)
+        {
+            uint16_t pattern = 0xdead;
+            memset_pattern4(gOut[j], &pattern, bufferSize);
```

Comment:
AddressSanitizer is reporting an overflow here.

Reply:
Done